pkg/storaged/btrfs/device.jsx

```diff
@@ -0,0 +1,87 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2023 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from "cockpit";
+import React from "react";
+import client from "../client";
+
+import { Button } from "@patternfly/react-core/dist/esm/components/Button/index.js";
+import { CardBody } from "@patternfly/react-core/dist/esm/components/Card/index.js";
+import { DescriptionList } from "@patternfly/react-core/dist/esm/components/DescriptionList/index.js";
+
+import { StorageCard, StorageDescription, new_card, register_crossref } from "../pages.jsx";
+import { StorageUsageBar } from "../storage-controls.jsx";
+import { btrfs_device_usage } from "./utils.jsx";
+
+const _ = cockpit.gettext;
+
+export function make_btrfs_device_card(next, backing_block, content_block) {
+    const block_btrfs = client.blocks_fsys_btrfs[content_block.path];
+    const label = block_btrfs && block_btrfs.data.label;
+    const uuid = block_btrfs && block_btrfs.data.uuid;
+    const use = btrfs_device_usage(client, uuid, block_btrfs.path);
+
+    const btrfs_card = new_card({
+        title: _("btrfs device"),
+        location: label || uuid,
+        next,
+        component: BtrfsDeviceCard,
+        props: { backing_block, content_block },
+    });
```

Comment:
This needs the standard actions, like "Unlock" and "Format".  That doesn't really make sense and we need to move those actions somewhere else, but that's the current mess we are in. Check any of the other card functions called by make_block_page.

Reply:
I added format, this works now. But I don't fully understand when `Unlock` happens. I tested with a XFS FS and that seems to mount fine with a passphrase.

Reply:
Aha! Locked is for when it's LUKS => and no entry in fstab. So our test matrix expands again:

* LUKS btrfs in fstab
* LUKS btrfs not in fstab

Reply:
Alright, so when we have a locked LUKS and in fstab, we can't do much for btrfs as there is limited info so this will have to do:

![image](https://github.com/cockpit-project/cockpit/assets/67428/2a8e50d9-3159-48ec-a320-0abd13209f2c)

Downside is that custom options are shown, but I don't know how we can avoid this honestly.
![image](https://github.com/cockpit-project/cockpit/assets/67428/27321364-c747-473c-90c3-d2a7a629b412)

Reply:
For a btrfs device it now shows:

![image](https://github.com/cockpit-project/cockpit/assets/67428/999b6c51-9243-4e0d-9d3b-287fa545ac2c)

Reply:
The locked action is handled by the encryption card when it's not in fstab. So concluding I would say this all works as expected:

![image](https://github.com/cockpit-project/cockpit/assets/67428/6204acf2-95f6-4305-9383-92dd9cb2fc6e)

Reply:
> Alright, so when we have a locked LUKS and in fstab, we can't do much for btrfs as there is limited info so this will have to do:

Hmm, if we say that a btrfs device is like a LVM2 physical volume, then ideally this row would show "Locked data" and not offer a "Mount" action. We could detect this by looking into the fstab options, and if there is subvol or subvolid in there, we treat it as unknown data.

If this _is_ a btrfs but without subvol/subvolid, then Cockpit doesn't know that it's btrfs and will offer to mount it, but that's ok in this special case I think.

There is a test in make_block_page like this:

```
if (fstab_config.length > 0) {
            card = make_filesystem_card(card, block, null, fstab_config);
} else {
            card = make_locked_encrypted_data_card(card, block);
}
```

This is the place where we would check the options and not call make_filesystem_card when finding subvol etc.

Reply:
I have pushed a commit that should improve this: https://github.com/cockpit-project/cockpit/pull/19690/commits/2d62db74f7221dc0b420904dcbb02629c19f99db